### PR TITLE
support callback request subscribers and job notification refactoring

### LINKED ISSUE #230: [Feature] Support and validate Job notification callback URL
## Description

Validate the passing of a subscriber-URL in an execute request.
- https://github.com/opengeospatial/ogcapi-processes/blob/master/openapi/schemas/processes-core/execute.yaml
- https://github.com/opengeospatial/ogcapi-processes/blob/master/core/abstract_tests/callback/ATS_process-execute-callback.adoc#L8

relates to #53 

Weaver supports `notification_email` from job execution, but could easily support the equivalent operation with the `subscriber` schema. https://github.com/opengeospatial/ogcapi-processes/blob/master/openapi/schemas/processes-core/subscriber.yaml
The corresponding job status requests should be set to relevant callback URLs the `subscriber` defined. 

## Related

- https://github.com/opengeospatial/PubSub-Whitepaper
  DRAFT report: https://github.com/opengeospatial/PubSub-Whitepaper/blob/master/20-046.pdf
- https://github.com/opengeospatial/pubsub

### PR DESCRIPTION
**Pull request by fmigneault**
## Changes

- Add `Job` ``subscribers`` support to define `OGC`-compliant callback URLs where HTTP(S) requests will be sent upon
  reaching certain `Job` status milestones (fixes #230).
- Add email notification support to the new ``subscribers`` definition (extension over `OGC` minimal requirements).
- Deprecate `Job` ``notification_email`` in the `OpenAPI` specification in favor of ``subscribers``, but preserve
  parsing of its value if provided in the `JSON` body during `Job` submission for backward compatibility support of
  existing servers. The ``Job.notification_email`` attribute is removed to avoid duplicate references.
- Add notification email for `Job` ``started`` status, only available through the ``subscribers`` property.
- Add ``{PROCESS_ID}/{STATUS}.mako`` template detection under the ``weaver.wps_email_notify_template_dir`` location
  to allow per-`Process` and per-`Job` status email customization.
- Refactor ``weaver/notify.py`` and ``weaver/processes/execution.py`` to avoid mixed references to the
  encryption/decryption logic employed for notification emails. All notifications including emails and
  callback requests are now completely handled and contained in the ``weaver/notify.py`` module.
- Remove partially duplicate Mako Template definition as hardcoded string and separate file for email notification.

## References

- Relates to https://github.com/opengeospatial/ogcapi-processes/issues/366
- Relates to #571
- Relates to #568 
- Fixes #230

### CODE REVIEW

**Review by perronld — DISMISSED** [review]
Works well with Finch!
( using notification_email)

File: tests/functional/test_cli.py [hunk 1]
```diff
@@ -566,6 +573,63 @@ def test_execute_with_auto_monitor(self):
...
+            assert mocked_requests.call_args_list[0].kwargs["json"]["status"] in running_statuses  # status JSON
+            assert mocked_requests.call_args_list[1].args == ("POST", subscribers["successUri"])
+            assert mocked_requests.call_args_list[1].kwargs["json"] == expect_outputs  # results JSON
+
+            # first argument None is 'from_addr' not configured, this is allowed if provided by 'From' email header
+            assert mocked_emails.call_count == 2, "Should not have sent both failed/success email notifications"
+            assert mocked_emails.call_args_list[0].args[:2] == (None, subscribers["inProgressEmail"])
+            assert b"Job test-client-Echo Started" in mocked_emails.call_args_list[0].args[-1]
+            assert mocked_emails.call_args_list[1].args[:2] == (None, subscribers["successEmail"])
+            assert b"Job test-client-Echo Succeeded" in mocked_emails.call_args_list[1].args[-1]
```
Comment (ChaamC) [inline tests/functional/test_cli.py:631]: Consider reusing variables for "test-client" / "Echo".

**Review by ChaamC — COMMENTED** [review]
Minor fixes on outputs/comments/variable names, the rest looks good to me. 
Good job!

File: weaver/notify.py [hunk 2]
```diff
@@ -178,3 +165,109 @@ def decrypt_email(email, settings):
...
+    For backward compatibility, ``notification_email`` directly provided at the root will be used if corresponding
+    definitions were not provided for the corresponding subscriber email fields.
+    """
+    notification_email = job_body.get("notification_email")
+    submit_subscribers = job_body.get("subscribers") or {}
+    mapped_subscribers = {}
+    for status, name, sub_type, alt in [
+        (Status.STARTED, "inProgressEmail", "emails", None),
+        (Status.FAILED, "failedEmail", "emails", notification_email),
+        (Status.SUCCEEDED, "successEmail", "emails", notification_email),
```
[the remark this answers is not in the document]
Comment (fmigneault) [reply weaver/notify.py:185]: Indeed. They are more "fallbacks" for backward compatibility of what `notification_email` accomplished before. There was no previous equivalent for callback requests and the emali on job start.

**Review by Nazim-crim — DISMISSED** [review]
LGTM, if I understood this correctly, multiple subscribers are only supported for different job status. Do you think it could be interesting to allow multiple emails notification for each status ?

**Comment by fmigneault** [comment]
> LGTM, if I understood this correctly, multiple subscribers are only supported for different job status. 

Yes 

> Do you think it could be interesting to allow multiple emails notification for each status ?

Could be, but I think we can work around it with a mailing list for the moment without changing the code. 
If an actual use case to support multiple emails arises, it could be added later on.

**Pull request merged**